### ISSUE TITLE
Middleware error Error: socket hang up

### ISSUE THREAD

**Issue by vilten** [lwsjs/local-web-server#124]
Hello,

OS: Centos 7
Node: 12.6.0
Local-web-server: 3.0.4

```
ws --key private-key.pem --cert lws-cert.pem -p 3094 --spa index.html --rewrite '/api/(.*) -> http://10.91.0.153:3090/api/$1' -z --verbose
```

HTTP request:

```
curl -ik 'https://localhost:3094/api/auth/login/' -H 'Accept: application/json' -H 'Referer: https://localhost:3094/' -H 'Origin: https://localhost:3094' -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36' -H 'Authorization: null' -H 'Content-Type: application/json; charset=UTF-8' --data-binary '{"username":"admin","password":"admin"}' --compressed
```

Log:

```
{
  'server.config': {
    key: <Buffer 2d 2d 2d 2d 2d 42 45 47 49 4e 20 52 53 41 20 50 52 49 56 41 54 45 20 4b 45 59 2d 2d 2d 2d 2d 0a 4d 49 49 45 70 41 49 42 41 41 4b 43 41 51 45 41 30 34 ... 1629 more bytes>,
    cert: <Buffer 2d 2d 2d 2d 2d 42 45 47 49 4e 20 43 45 52 54 49 46 49 43 41 54 45 2d 2d 2d 2d 2d 0a 4d 49 49 44 51 44 43 43 41 69 69 67 41 77 49 42 41 67 49 4a 41 4c ... 1138 more bytes>
  }
}
{ 'middleware.cors.config': {} }
{ 'middleware.compress.config': { compress: true } }
{
  'middleware.rewrite.config': {
    rewrite: [ { from: '/api/(.*)', to: 'http://10.91.0.153:3090/api/$1' } ]
  }
}
{ 'middleware.conditional-get.config': {} }
{
  'middleware.spa.config': {
    spa: 'index.html',
    root: '/opt/vilten/encoder/enc_gui',
    spaAssetTest: undefined,
    spaAssetTestFs: undefined
  }
}
{
  'middleware.static.config': { hidden: true, root: '/opt/vilten/encoder/enc_gui' }
}
{ 'middleware.index.config': { icons: true } }
Listening on https://localhost.localdomain:3094, https://127.0.0.1:3094, https://10.91.0.153:3094
{
  'server.socket.new': { bytesRead: '0', bytesWritten: '0', remoteAddress: '::1' }
}
{
  'server.request': {
    requestId: 1,
    method: 'POST',
    url: '/api/auth/login/',
    headers: {
      host: 'localhost:3094',
      'accept-encoding': 'deflate, gzip',
      accept: 'application/json',
      referer: 'https://localhost:3094/',
      origin: 'https://localhost:3094',
      'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36',
      authorization: 'null',
      'content-type': 'application/json; charset=UTF-8',
      'content-length': '39'
    },
    body: { username: 'admin', password: 'admin' }
  }
}
{
  'middleware.rewrite.remote.request': {
    rewrite: {
      id: 1,
      from: '/api/auth/login/',
      to: 'http://10.91.0.153:3090/api/auth/login/'
    },
    method: 'POST',
    headers: {
      host: '10.91.0.153:3090',
      'accept-encoding': 'deflate, gzip',
      accept: 'application/json',
      referer: 'https://localhost:3094/',
      origin: 'https://localhost:3094',
      'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36',
      authorization: 'null',
      'content-type': 'application/json; charset=UTF-8',
      'content-length': '39'
    }
  }
}
{
  'process.memoryUsage': {
    rss: '34.2 MB',
    heapTotal: '11.9 MB',
    heapUsed: '9.4 MB',
    external: '1.3 MB'
  }
}
{
  'process.memoryUsage': {
    rss: '34.2 MB',
    heapTotal: '11.9 MB',
    heapUsed: '9.5 MB',
    external: '1.3 MB'
  }
}
{
  'server.socket.close': { bytesRead: '753 B', bytesWritten: '1.3 kB', remoteAddress: '::1' }
}
5:32:37 PM: Middleware error
Error: socket hang up
    at connResetException (internal/errors.js:559:14)
    at Socket.socketOnEnd (_http_client.js:436:23)
    at Socket.emit (events.js:208:15)
    at endReadableNT (_stream_readable.js:1154:12)
    at processTicksAndRejections (internal/process/task_queues.js:77:11)
```

On MacOSx with same versions of local-web-server and nodejs is everything OK, on CentOS it hangs out.

Thanx for your help

**Comment by 75lb** [comment]
At first glance, it looked like your `10.91.0.153` server rejected this incoming request:

```
{
  'middleware.rewrite.remote.request': {
    rewrite: {
      id: 1,
      from: '/api/auth/login/',
      to: 'http://10.91.0.153:3090/api/auth/login/'
    },
    method: 'POST',
    headers: {
      host: '10.91.0.153:3090',
      'accept-encoding': 'deflate, gzip',
      accept: 'application/json',
      referer: 'https://localhost:3094/',
      origin: 'https://localhost:3094',
      'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36',
      authorization: 'null',
      'content-type': 'application/json; charset=UTF-8',
      'content-length': '39'
    }
  }
}
```

Could you compare this `middleware.rewrite.remote.request` with the same request on your mac? For some reason, it let your mac request through and my the difference between these two remote requests will reveal why.

**Comment by 75lb** [comment]
does it work on both Mac and Linux with node v10?

**Comment by vilten** [comment]
Hello,

10.91.0.153 is only local IP, curl in this IP directly without proxy through local-web-server works correctly. I tried it also with localhost or 127.0.0.1 with the same result - hang out.

Should I try it with node v10? For some specific reasons?

**Comment by 75lb** [comment]
What is the difference between your mac and linux `--verbose` output? Could you send me the mac verbose output so I can see the differences? 

Node v12 has been a buggy release, i have reported three issues already. Test with v10 if you can, it's easy using `nvm`.

**Comment by 75lb** [comment]
did you resolve this issue?

**Comment by vilten** [comment]
Sorry for delay, ... a lot of work on another projects. I will try it today.

> did you resolve this issue?

**Comment by holiakho** [comment]
Hi, same issue:
What I have:
lws of 3.0.4 running in docker node:10.16.0-alpine 
`ws --rewrite "/api/(.*) -> http://192.168.2.121:4000/api/$1" --spa index.html`

curl from docker container works fine, but when hosted app trying to POST - it waits some time and fail with

    Error: socket hang up
    at createHangUpError (_http_client.js:323:15)
    at Socket.socketOnEnd (_http_client.js:426:23)
    at Socket.emit (events.js:203:15)
    at endReadableNT (_stream_readable.js:1129:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)`

after downgrading to 2.6.0 I was able to reach server, but requests were not proxied(rewrite doesn't work)

**Comment by 75lb** [comment]
Hi, this looks like a duplicate of #127 - please follow that ticket. 👍

**Comment by 75lb** [comment]
Fixed in lws-rewrite v2.0.6. Please reinstall local-web-server and verify whether it now works correctly for you. Thanks.
